### PR TITLE
Implement Symbol.prototype.description

### PR DESCRIPTION
**Pull request by szilagyiadam**
The algorithm is based on ECMA-262 v11, 19.4.3.2

Also added a custom dispatcher for the Symbol prototype

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c [hunk 1]
```diff
@@ -89,6 +89,47 @@ ecma_builtin_symbol_prototype_object_to_primitive (ecma_value_t this_arg) /**< t
...
+ * See also:
+ *          ECMA-262 v11, 19.4.3.2
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_symbol_prototype_description_getter (ecma_value_t this_arg) /**< this argument */
+{
+  if (ecma_is_value_symbol (this_arg))
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c:104]: There is already a `ecma_symbol_to_string_helper` function which almost do the same. The second parameter of this function currently is boolean to decide whether `toString` or `valueOf` routine should be performed. Please change this boolean into an enum, and add `description` routine as well into this list. After that please modify this helper function according to it.

**Comment by szilagyiadam** [comment]
I've made a custom dispatcher for the Symbol.prototype, since the 4 existing methods (valueOf, toString, toPrimitive, description) are very similar, and this way the code is more readable IMHO

File: jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c [hunk 2]
```diff
@@ -45,49 +62,44 @@
...
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_symbol_prototype_object_to_primitive (ecma_value_t this_arg) /**< this argument */
-{
-  return ecma_builtin_symbol_prototype_object_value_of (this_arg);
-} /* ecma_builtin_symbol_prototype_object_to_primitive */
+  ecma_value_t sym = ecma_symbol_this_value (this_arg);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c:85]: This `ecma_symbol_this_value` call should be performed only once.

File: jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c [hunk 3]
```diff
@@ -45,49 +62,44 @@
...
-ecma_builtin_symbol_prototype_object_to_primitive (ecma_value_t this_arg) /**< this argument */
-{
-  return ecma_builtin_symbol_prototype_object_value_of (this_arg);
-} /* ecma_builtin_symbol_prototype_object_to_primitive */
+  if (builtin_routine_id < ECMA_SYMBOL_PROTOTYPE_TO_STRING)
+  {
+    return ecma_copy_value (sym);
+  }
+
+  if (ECMA_IS_VALUE_ERROR (sym))
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c:87]: This check should be before `if (builtin_routine_id < ECMA_SYMBOL_PROTOTYPE_TO_STRING)`

File: jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c [hunk 4]
```diff
@@ -45,49 +62,44 @@
...
+  if (builtin_routine_id == ECMA_SYMBOL_PROTOTYPE_TO_STRING)
+  {
+    return ecma_get_symbol_descriptive_string (sym);
+  }
+
+  JERRY_ASSERT (builtin_routine_id == ECMA_SYMBOL_PROTOTYPE_DESCRIPTION);
+  ecma_string_t *symbol_p = ecma_get_symbol_from_value (sym);
+  ecma_string_t *desc_p = ecma_get_symbol_description (symbol_p);
+  ecma_ref_ecma_string (desc_p);
+  return ecma_make_string_value (desc_p);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c:101]: new line before.

File: jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c [hunk 5]
```diff
@@ -30,6 +30,23 @@
...
+ * This object has a custom dispatch function.
+ */
+#define BUILTIN_CUSTOM_DISPATCH
+
+/**
+ * List of built-in routine identifiers.
+ */
+enum
+{
+  ECMA_Symbol_PROTOTYPE_ROUTINE_START = ECMA_BUILTIN_ID__COUNT - 1,
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-symbol-prototype.c:43]: This entry seems to have incorrect casing, or maybe not?

File: jerry-core/ecma/operations/ecma-symbol-object.c [hunk 6]
```diff
@@ -137,42 +137,42 @@ ecma_get_symbol_descriptive_string (ecma_value_t symbol_value) /**< symbol to st
...
       {
-        return (is_to_string ? ecma_get_symbol_descriptive_string (ext_object_p->u.class_prop.u.value)
-                             : ecma_copy_value (ext_object_p->u.class_prop.u.value));
+        return ext_obj_p->u.class_prop.u.value;
       }
     }
   }
 
+  /* 3. */
   return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is must be a Symbol."));
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-symbol-object.c:174]: If we are modifying around this then the error message can be also fixed. The `is` in the sentence feels awkward imho.

**Pull request merged**